src/main/java/sx/blah/discord/util/components/ComponentRegistry.java

```diff
@@ -0,0 +1,227 @@
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This is a central registry for components. Use this if you need components in a dynamic setting (i.e.
+ * the {@link ComponentInjection} annotation cannot be used for a particular use case or you are not using the module
+ * system in the first place).
+ */
+public class ComponentRegistry {
+	
+	private final Map<Class<? extends IComponent>, IComponentProvider<? extends IComponent>> providerMap = new ConcurrentHashMap<>();
+	
+	private static final MethodHandles.Lookup lookup = MethodHandles.lookup();
+	
+	public ComponentRegistry() {}
+	
+	/**
+	 * This registers an {@link IComponentProvider} with the component registry.
+	 *
+	 * @param provider The provider to register.
+	 */
+	public void registerComponentProvider(IComponentProvider<?> provider) {
+		if (providerMap.containsKey(provider.getComponentClass())) {
+			throw new IllegalStateException("Cannot register provider for class " + provider.getComponentClass() + ", the provider for it already exists!");
+		}
+		
+		providerMap.put(provider.getComponentClass(), provider);
+	}
+	
+	/**
+	 * Gets all the registered component providers.
+	 *
+	 * @return The component providers.
+	 */
+	public Collection<IComponentProvider<? extends IComponent>> getComponentProviders() {
+		return providerMap.values();
+	}
+	
+	/**
+	 * Gets all the classes representing possible components which may be injected.
+	 *
+	 * @return The possible component classes.
+	 */
+	public Collection<Class<? extends IComponent>> getComponentClasses() {
+		return providerMap.keySet();
+	}
+	
+	/**
+	 * This performs automatic dependency injection into an object. This means that the object must contain
+	 * {@link ComponentInjection} annotations in order for this to have any effect.
+	 *
+	 * @param obj The object to perform component injection into.
+	 */
+	public void injectInto(Object obj) {
+		new Injector(obj).inject();
+	}
+	
+	/**
+	 * This creates an {@link Injector} instance for the provided object.
+	 *
+	 * @param obj The object to perform injection on.
+	 * @return The injector.
+	 */
+	public Injector injectorFor(Object obj) {
+		return new Injector(obj);
+	}
+	
+	/**
+	 * This class is used to perform component injection in an object.
+	 */
+	public class Injector {
+		
+		private final Object parent;
+		
+		private Injector(Object parent) {
+			this.parent = parent;
+		}
+		
+		/**
+		 * Performs automated component injection into the object.
+		 *
+		 * @see #injectInto(Object)
+		 */
+		public void inject() {
+			//Inject into fields
+			FieldUtils.getFieldsListWithAnnotation(parent.getClass(), ComponentInjection.class)
+					.stream()
+					.peek(f -> f.setAccessible(true))
+					.peek(FieldUtils::removeFinalModifier)
+					.forEach(field -> {
+						try {
+							field.set(parent, createComponent((Class<? extends IComponent>) Class.forName(field.getAnnotation(ComponentInjection.class).value())));
```

Comment:
Class should be discovered or registered elsewhere (instead of annotation value) to avoid coupling the injectable object with the concrete impl of the IComponent to inject.